Python/specialize.c

```diff
@@ -2416,13 +2416,24 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
 
 /* float-long */
 
-static int
+// Guards should check that the float part is not NaN.
+// Guards for / (aka true_div) should check for 0 or 0.0 as the rhs.
+static inline int
 float_compactlong_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
         PyFloat_CheckExact(lhs) &&
         PyLong_CheckExact(rhs) &&
-        _PyLong_IsCompact((PyLongObject *)rhs)
+        _PyLong_IsCompact((PyLongObject *)rhs) &&
+        !isnan(PyFloat_AsDouble(lhs))
```

Comment:
I'd do all the checks for las first and then the checks for rhs. Easier to read.